compiler/rustc_hir_typeck/src/writeback.rs

```diff
@@ -363,9 +363,10 @@ impl<'cx, 'tcx> Visitor<'tcx> for WritebackCx<'cx, 'tcx> {
 
     fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty<'tcx>) {
         intravisit::walk_ty(self, hir_ty);
-        let ty = self.fcx.node_ty(hir_ty.hir_id);
-        let ty = self.resolve(ty, &hir_ty.span);
-        self.write_ty_to_typeck_results(hir_ty.hir_id, ty);
+        if let Some(ty) = self.fcx.node_ty_opt(hir_ty.hir_id) {
```

Comment:
From what I remember, this is not a correct fix, it hides the underlying issue.
Nodes should indeed always have types (possibly erroneous) attached to them, and `fcx.node_ty_opt` should always return `Some` here.

Reply:
Currently I find the root cause is because the generic arguments are empty for code `S<impl Oops>`.
so we return at here:
https://github.com/rust-lang/rust/blob/5e6de2369c82ed0b36e6b651b041bad5cb5e1ef8/compiler/rustc_hir_analysis/src/astconv/mod.rs#L367

So we didn't call write_ty for this hir_id.

I think we can skip the generic arguments part with error in the phase of writeback.
@petrochenkov How do you think on current fix？